homeassistant/components/sensor/__init__.py

```diff
@@ -273,6 +277,61 @@ def state(self) -> Any:
         """Return the state of the sensor and perform unit conversions, if needed."""
         unit_of_measurement = self.native_unit_of_measurement
         value = self.native_value
+        device_class = self.device_class
+
+        # We have an old non-datetime value, warn about it and convert it during
+        # the deprecation period.
+        if (
+            value is not None
+            and device_class in (DEVICE_CLASS_DATE, DEVICE_CLASS_TIMESTAMP)
+            and not isinstance(value, (date, datetime))
+        ):
+            # Deprecation warning for date/timestamp device classes
+            if not self.__datetime_as_string_deprecation_logged:
+                report_issue = self._suggest_report_issue()
+                _LOGGER.warning(
+                    "%s is providing a string for its state, while the device "
+                    "class is '%s', this is not valid and will be unsupported "
+                    "from Home Assistant 2022.2. Please %s",
+                    self.entity_id,
+                    device_class,
+                    report_issue,
+                )
+                self.__datetime_as_string_deprecation_logged = True
+
+            # Anyways, lets validate the date at least..
+            try:
+                value = ciso8601.parse_datetime(str(value))
+            except (ValueError, IndexError) as error:
+                raise ValueError(
+                    f"Invalid date/datetime: {self.entity_id} provide state '{value}', "
+                    f"while it has device class '{device_class}'"
+                ) from error
+
+            # Convert the date object to a standardized state string.
+            if device_class == DEVICE_CLASS_DATE:
+                return value.date().isoformat()
+            return value.isoformat(timespec="seconds")
+
+        # Received a datetime
+        if value is not None and device_class == DEVICE_CLASS_TIMESTAMP:
+            try:
+                return value.isoformat(timespec="seconds")  # type: ignore
+            except (AttributeError, TypeError) as err:
+                raise ValueError(
```

Comment:
Should we add tests for these error cases?

Reply:
We can't at this point, I've removed them in `e07facc`, as with the current method, they are snatched by the backward compatibility layer above. We can transform the tests from the backward compatibility once that is removed.